ZenPacks/zenoss/OpenStackInfrastructure/model_mappers.py

```diff
@@ -0,0 +1,1045 @@
+    return hypervisors
+
+
+def map_servers(device, results, hostmap):
+    '''Instance maps'''
+
+    servers = []
+    for server in results['servers']:
+        if not server.get('id', None):
+            continue
+
+        # Backup support is optional. Guard against it not existing.
+        backup_schedule_enabled = server.get('backup_schedule',
+                                             {}).get('enabled', False)
+        backup_schedule_daily = server.get('backup_schedule',
+                                           {}).get('daily', 'DISABLED')
+        backup_schedule_weekly = server.get('backup_schedule',
+                                            {}).get('weekly', 'DISABLED')
+
+        # Get and classify IP addresses into public and private: (fixed/floating)
+        public_ips = set()
+        private_ips = set()
+
+        access_ipv4 = server.get('accessIPv4')
+        if access_ipv4:
+            public_ips.add(access_ipv4)
+
+        access_ipv6 = server.get('accessIPv6')
+        if access_ipv6:
+            public_ips.add(access_ipv6)
+
+        address_group = server.get('addresses')
+        if address_group:
+            for network_name, net_addresses in address_group.items():
+                for address in net_addresses:
+                    if address.get('OS-EXT-IPS:type') == 'fixed':
+                        private_ips.add(address.get('addr'))
+                    elif address.get('OS-EXT-IPS:type') == 'floating':
+                        public_ips.add(address.get('addr'))
+                    else:
+                        LOG.info("Address type not found for %s", address.get('addr'))
+                        LOG.info("Adding %s to private_ips", address.get('addr'))
+                        private_ips.add(address.get('addr'))
+
+        # Flavor and Image IDs could be specified two different ways.
+        flavor_id = server.get('flavorId', None) or \
+            server.get('flavor', {}).get('id', None)
+
+        tenant_id = server.get('tenant_id', '')
+
+        power_state = server.get('OS-EXT-STS:power_state', 0)
+        task_state = server.get('OS-EXT-STS:task_state')
+        if not task_state:
+            task_state = 'no task in progress'
+        vm_state = server.get('OS-EXT-STS:vm_state')
+
+        # Note: volume relations are added in volumes map below
+        server_dict = dict(
+            id=prepId('server-{0}'.format(server['id'])),
+            title=server.get('name', server['id']),
+            resourceId=server['id'],
+            serverId=server['id'],  # 847424
+            serverStatus=server.get('status', '').lower(),
+            serverBackupEnabled=backup_schedule_enabled,
+            serverBackupDaily=backup_schedule_daily,
+            serverBackupWeekly=backup_schedule_weekly,
+            publicIps=list(public_ips),
+            privateIps=list(private_ips),
+            set_flavor=prepId('flavor-{0}'.format(flavor_id)),
```

Comment:
I'm going to comment here, but really this is a general thing.  One part of making this modeler tolerate missing data is thinking about what happens if these IDs don't exist- that is, we reference a flavor here by ID that we couldn't find when we listed out the flavors in map_flavors.

For now, we can probably ignore this issue- it will generate warnings in the zenhub log when these setters try to find the target object by ID and fail.   But it would be prudent to add code to the main process_method that scans through all objectmaps to build a list of all valid IDs, then looks for calls to set_foo that reference something not in that list.   I think I've done similar things before (probably vsphere) and it's not that hard.. it lets us put in warnings and catch possible bugs, and to strip the set calls that we know will generate zenhub errors.   So i think it's worth doing, if not critical.

Reply:
At that point, it may be worth looking at Che's new modeler which can tolerate missing pieces. That is a significant amount of work to implement, but it may be worth it if we invest a lot of time doing other refactoring.